contracts/ICS20Transfer.sol

```diff
@@ -23,25 +24,52 @@ using SafeERC20 for IERC20;
  * - Separate escrow balance tracking
  * - Related to escrow ^: invariant checking (where to implement that?)
  */
-contract ICS20Transfer is IIBCApp, IICS20Transfer, IICS20Errors, Ownable, ReentrancyGuardTransient, Multicall {
-    /// @notice The escrow contract address
-    IEscrow private immutable ESCROW;
-    /// @notice Mapping of non-native denoms to their respective IBCERC20 contracts created here
-    mapping(string denom => IBCERC20 ibcERC20Contract) private _ibcDenomContracts;
-
-    /// @param owner_ The owner of the contract
-    constructor(address owner_) Ownable(owner_) {
-        ESCROW = new Escrow(address(this));
+contract ICS20Transfer is
+    IIBCApp,
+    IICS20Transfer,
+    IICS20Errors,
+    Initializable,
+    Ownable,
+    ReentrancyGuardTransient,
+    Multicall
+{
+    /// @notice Storage of the ICS20Transfer contract
+    /// @dev It's implemented on a custom ERC-7201 namespace to reduce the
+    /// @dev risk of storage collisions when using with upgradeable contracts.
+    /// @param escrow The escrow contract
+    /// @param ibcDenomContracts Mapping of non-native denoms to their respective IBCERC20 contracts
+    /// @custom:storage-location erc7201:ibc.storage.ICS20Transfer
+    struct ICS20TransferStorage {
```

Comment:
I'm curious if this has any penalty in gas cost? Would this end up fetching more data than we need (+ use more memory) whenever we need only one of them?

I hope that the fact that it is set as a storage type helps, but still want to make sure we are not getting punished too hard for this :)

Have you tried to run the benchmarks to see if anything has changed significantly?

Reply:
Based on my understanding of solidity, this doesn't fetch more data/memory than we need, but we might be consuming more gas since we are making some extra function calls.

I updated the benchmarks and there is an increase in gas usage, but this might be due to the addition of the proxy contract, and fallback logic as well. The gas increase is not substantial in the aggregated scenario.